eval/topdown_test.go

```diff
@@ -423,10 +432,111 @@ func TestTopDownEqExpr(t *testing.T) {
 		case string:
 			expected := loadExpectedResult(e)
 			result, err := TopDownQuery(ctx)
+
+			switch ctx.Rule.DocKind() {
+			case opalog.PartialSetDoc:
+				sort.Sort(ResultSet(result.([]interface{})))
+			}
+
+			if err != nil {
+				t.Errorf("Test case %d (%v): unexpected error: %v", i+1, tc.note, err)
+				continue
+			}
+			if !reflect.DeepEqual(result, expected) {
+				t.Errorf("Test case %d (%v): expected %v but got: %v", i+1, tc.note, expected, result)
+			}
+		}
+	}
+
+}
+
+func TestTopDownVirtualDocs(t *testing.T) {
+
+	tests := []struct {
+		note     string
+		rules    []string
+		expected interface{}
+	}{
+		// input to partial set and object docs
+		{"input: set 1", []string{"p = true :- q[1]", "q[x] :- a[i] = x"}, "true"},
+		{"input: set 2", []string{"p[x] :- q[1] = x", "q[x] :- a[i] = x"}, "[true]"},
+		{"input: set embedded", []string{`p[x] :- x = {"b": [q[2]]}`, `q[x] :- a[i] = x`}, `[{"b": [true]}]`},
+		{"input: set undefined", []string{"p = true :- q[1000]", "q[x] :- a[x] = y"}, ""},
+		{"input: object 1", []string{"p = true :- q[1] = 2", "q[i] = x :- a[i] = x"}, "true"},
+		{"input: object 2", []string{"p = true :- q[1] = 0", "q[x] = i :- a[i] = x"}, "true"},
+		{"input: object embedded 1", []string{"p[x] :- x = [1, q[3], q[2]]", "q[i] = x :- a[i] = x"}, "[[1,4,3]]"},
+		{"input: object embedded 2", []string{`p[x] :- x = {"a": [q[3]], "b": [q[2]]}`, `q[i] = x :- a[i] = x`}, `[{"a": [4], "b": [3]}]`},
+		{"input: object undefined val", []string{`p = true :- q[1] = 9999`, `q[i] = x :- a[i] = x`}, ""},
+		{"input: object undefined key 1", []string{`p = true :- q[9999] = 2`, `q[i] = x :- a[i] = x`}, ""},
+		{"input: object undefined key 2", []string{`p = true :- q["foo"] = 2`, `q[i] = x :- a[i] = x`}, ""},
+		{"input: object dereference ground", []string{`p = true :- q[0]["x"][1] = false`, `q[i] = x :- x = c[i]`}, "true"},
+		{"input: object defererence non-ground", []string{`p = true :- q[0][x][y] = false`, `q[i] = x :- x = c[i]`}, "true"},
+
```

Comment:
Not sure if these are here, but I haven't found them....
// virtual docs on both sides of the equality...
p[x] :- vdoc1[x] = vdoc2[x] 

// walking over a list of values for a virtual object, where some of the keys are undefined
p[y] :- vdoc1[x], vdoc2[x] = y